### PR TITLE
Separate scripts and reports

### PR DESCRIPTION
**Pull request by jonasfreimuth**
Additional changes
* remove sigmut db
* format scripts when touched
* update test/settings.yaml with current database locations
* convert file io to use CSV files wherever possible

### CODE REVIEW

**Comment by jonasfreimuth** [comment]
Will be set ready for review once pipeline run of NY_merge data nets same results as the [report](https://bimsbstatic.mdc-berlin.de/akalin/AAkalin_pathogenomics/sarscov2_ww_reports/220225_dataset_NYC_RBD/index.html) linked in the current (v3) preprint of the associated paper.

**Comment by jonasfreimuth** [comment]
`make distcheck` completes successfully.

File: scripts/Sample_prep_scripts/sigmuts_on_amplicons.rmd [hunk 1]
```diff
@@ -34,7 +34,7 @@ amplicon_gr <- makeGRangesFromDataFrame ( amplicon, keep.extra.columns = TRUE )
 ```{r}
 # TODO: make the mutation sheet with nt position for ALL mutatation and loop through it
 # mut <- read.csv("../tests/mutation_sheet_detailed.csv", header = TRUE ) # TODO provide as input
-mut <- read.table("../tests/sample_data/signature_mutations/B117_outbreakinfo_mutation_report_data_2021-03-06.tsv", header = TRUE)
+mut <- read.csv("../tests/sample_data/signature_mutations/B117_outbreakinfo_mutation_report_data_2021-03-06.tsv", header = TRUE)
```
Comment (alexg9010) [inline scripts/Sample_prep_scripts/sigmuts_on_amplicons.rmd]: Why would you use the `read.csv` function to read a tab-separated file?
I would either add the right parameters (separators, headings) to the `read.table` call or change to `read.delim`, which is the function to read tsv files.

File: scripts/creating_mutation_summary_table.R [hunk 2]
```diff
@@ -28,14 +31,14 @@ create_summary <- function ( files, output_file ){
                             x = mutations_list)
 
   return(merged_mutations)
-
 }
 
 args <- commandArgs (trailingOnly=TRUE)
+#args <- c("/home/jfreige/proj/pigx_sars-cov-2/tests/output/mutations", "/home/jfreige/proj/pigx_sars-cov-2/tests/output/variants/data_mutation_plot.csv")
```
Comment (alexg9010) [inline scripts/creating_mutation_summary_table.R]: This should not be part of the PR.

File: scripts/creating_variants_summary_table.R [hunk 3]
```diff
@@ -13,12 +14,15 @@ create_summary <- function ( files ){
...
-                          FUN = read.table,
-                          sep = "\t",
+                          FUN = read.csv,
                           header = TRUE,
                           colClasses = "character",
                           check.names = FALSE )
 
+  # remove empty files from list
+  variants_list_has_rows <- sapply(variants_list, nrow)
+  variants_list <- variants_list[variants_list_has_rows > 0]
```
Comment (alexg9010) [inline scripts/creating_variants_summary_table.R]: If you don't reuse the `variants_list_has_rows` elsewhere, I suggest you combine the two lines into one.

File: scripts/fun_index.R [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  ))
+}
+
+# function to dump data in csv formatted string
+dump_csv <- function(data, format = "csv", rownames = FALSE) {
+  if (format != "csv") stop("Data can only be dumped in CSV format.")
+  capture.output(write.csv(data,
+    file = "",
+    quote = FALSE, row.names = rownames
+  )) %>%
```
Comment (alexg9010) [inline scripts/fun_index.R:23]: To use the pipe you also need to load `library(dplyr)` here, or you rewrite this function without the pipe.

File: scripts/creating_variants_summary_table.R [hunk 5]
```diff
@@ -1,41 +1,40 @@
...
 
-args <- commandArgs (trailingOnly=TRUE)
-variants_dir <- args[1]
-output_file <- args[2]
+args <- commandArgs(trailingOnly = TRUE)
+
+output_file <- args[1]
+files       <- args[2:length(args)]
+
+output <- create_summary(files)
```
Comment (alexg9010) [inline scripts/creating_variants_summary_table.R]: The whole processing within the scripts `creating_variants_summary_table.R` and `creating_mutation_summary_table.R` is identical, except for suffices variants vs mutations. 
Since the input files and output filename are given by arguments, it might be a good idea to join both scripts into a `creating_variants_summary_table.R` script?

File: scripts/deconvolution_funs.R [hunk 6]
```diff
@@ -0,0 +1,255 @@
+library("stringr")
+library("dplyr")
+library("data.table")
+
+parse_snv_csv <- function(snvfile, ...) { # allele frequency from v-pipe vcf
```
Comment (alexg9010) [inline scripts/deconvolution_funs.R]: why do you need these `...` as arguments? They are not passed to any other downstream function ...

File: scripts/deconvolution_funs.R [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  return(rlm_coefficients)
+  }
+
+deconv_debug <- function(bulk, sig) {
+  #' This function performs the deconvolution using a signature matrix for the
+  #' mutations found in the sample and bulk frequency values derived by the SNV
+  #' caller
+  #' it was build by Altuna
+
+  rlm_model <- suppressWarnings(MASS::rlm(sig, bulk, maxit = 100, method = "M"))
```
Comment (alexg9010) [inline scripts/deconvolution_funs.R:205]: Is the `MASS` R-library listed as a dependency?

File: scripts/pred_mutation_increase.R [hunk 8]
```diff
@@ -1,84 +1,45 @@
...
+      # (if not, this will return NULL)
+      if (sum(mutation_col > 0) >= 0.2 * length(dates)) {
+        lm      <- lm(mutation_col ~ dates)
+        lm_sum  <- summary(lm)
+
+        data.frame(
+          mutation     = mutation,
+          coefficients = lm_sum$coefficients["dates", "Estimate"],
+          pvalues      = lm_sum$coefficients["dates", "Pr(>|t|)"]
+        )
```
Comment (alexg9010) [inline scripts/pred_mutation_increase.R:35]: Nice job summarizing this code.  👍 
I hope you verified that the output is identical to the previous code.

File: snakefile.py [hunk 9]
```diff
@@ -37,14 +37,14 @@
...
 KRONA_DB         = config['locations']['krona-db-dir']
-SIGMUT_DB        = config['locations']['sigmut-db-dir']
 VEP_DB           = config['locations']['vep-db-dir']
 OUTPUT_DIR       = config['locations']['output-dir']
 
 # TODO: get default read length from multiqc
 READ_LENGTH      = config['trimming']['read-length']
 CUT_OFF          = config['trimming']['cut-off']
 
+MUTATION_DEPTH_THRESHOLD    = config["reporting"]["mutation-depth-threshold"]
```
Comment (alexg9010) [inline snakefile.py:47]: I saw where this is being used, but what makes it different from `MUTATION_COVERAGE_THRESHOLD`?

File: scripts/report_scripts/variantreport_p_sample.Rmd [hunk 10]
```diff
@@ -27,468 +25,182 @@ params:
...
+the sample and the deconvoluted lineage abundances. The tables below display the
+summarized results from single-nucleotide-variant (SNV) calling and their 
+translation into protein mutations. A bar chart visualizes the proportions of 
+lineages that are derived by deconvolution of the set of mutations that is 
+matching the provided list of lineage - characterizing mutations (signature 
+mutations). For detailed method description please consult the
+[documentation](http://bioinformatics.mdc-berlin.de/pigx_docs/pigx-sars-cov-2.html#output-description)
+*Note: Deconvolution results are biased by the number of reads and by the number
+of signature mutations that were found for this sample.  Check with the QC 
+report to be able to assess how trustworthy the results are.*
```
Comment (alexg9010) [inline scripts/report_scripts/variantreport_p_sample.Rmd:47]: Is this sentence still rendered as italic?

File: snakefile.py (hunk 9 above)
Comment (jonasfreimuth) [reply snakefile.py:47]: MUTATION_COVERAGE_THRESHOLD gives the percentage of mutations in the mutation_sheet that need to be covered by each sample to be used in the mutation regression, whereas MUTATION_DEPTH_THRESHOLD gives the minumum read coverage at the position of each mutation that needs to be met in order for the mutation to be considered as a signature mutation... I suppose I should mention that somewhere...

File: scripts/deconvolution_funs.R (hunk 7 above)
Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R:205]: Well, according to Dirk Eddelbuettel in 2024/16 (https://stackoverflow.com/a/9705725) it is at least in the recommended packages. But to be safe I will explicitly load it and include it in the manifest and other build files...

File: scripts/pred_mutation_increase.R (hunk 8 above)
Comment (jonasfreimuth) [reply scripts/pred_mutation_increase.R:35]: I have verified that the function itself outputs exactly the same object when running on the `sepjan_new` "benchmark" dataset as the previous version.

File: scripts/report_scripts/variantreport_p_sample.Rmd (hunk 10 above)
Comment (jonasfreimuth) [reply scripts/report_scripts/variantreport_p_sample.Rmd:47]: It is. Only the markup in the code looks different. But apparently as long as there is a `*` next to the last non whitespace character of the block it is fine...

File: scripts/fun_index.R (hunk 4 above)
Comment (jonasfreimuth) [reply scripts/fun_index.R:23]: package should be loaded now

File: scripts/creating_variants_summary_table.R (hunk 3 above)
Comment (jonasfreimuth) [reply scripts/creating_variants_summary_table.R]: Done

File: scripts/creating_variants_summary_table.R (hunk 5 above)
Comment (jonasfreimuth) [reply scripts/creating_variants_summary_table.R]: Done

File: scripts/deconvolution_funs.R (hunk 6 above)
Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R]: Removed it

File: scripts/deconvolution_funs.R (hunk 7 above)
Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R:205]: See the lines [here ](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R69), [here](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-f2d3269a9b49cddd2b846cf38ebbae68a5d7bbadb390668e38c5c3a773b599d1R1), and [here](https://github.com/BIMSBbioinfo/pigx_sars-cov-2/pull/121/files#diff-4a3434b9932613220180f203e1970bf52b99ab6a4d66fc0bcdde9e8a8060bc68R66)

File: scripts/deconvolution_funs.R [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    # e.g. Gene:ENSSASG00005000002 <> Gene:ENSSASG00005000003
+    distinct()
+
+    # TODO Previously deletions were handled by generating pseudo rows for each
+    # one. I currently do not have access to a file with deletions, and so
+    # cannot understand how to handle this properly. However this solution here
+    # is still better, because due to a bug the previous solution could almost
+    # never run. Now there is at least a warning...
+    if (any(!is.na(locations$mut_end))) {
+      warning("Found a deletion. Help.")
```
Comment (jonasfreimuth) [inline scripts/deconvolution_funs.R]: The dangling issue of how to deal with deletions here is fixed on my local branch btw. I will push once the review is through, or I am prompted to do so.

Comment (jonasfreimuth) [reply scripts/deconvolution_funs.R]: Oops, during some reshuffling of my local repo I didn't realize this PR would be affected and pushed onto the branch. The result should be that commit "simplify deconvolution" now has additional subcommits relating to this point, see below.
```
  remove unused detectable_deletions function

    I looked through the entire results section of the project dir
    and could find no files with "deletions" in them...

  remove calculation of lowecase others col

    This was supposed to already be gone before I worked on this.
    It was commented out, but when I copied stuff around I didn't
    realize it and activated it again.

  format deconvolution_funs.R

  add proper handling of indels to get_protein_mut

    This replaces the detectable_deletions function and works both for
    insertions and deletions. Plus it actually runs.

  fix variable in deconvolution_funs.R

    Missed it during renaming.

  update regex for nuc muts to include indels

    It is conjecture whether this works, but if amino acids have - for
    deletions, nucleotides probably do too.

  remove browser() debugging call i forgot

  fix concat of replacement rows in get_protein_mut

    Missed this earlier, didnt even test it then.

  fix deletion of indel rows in get_protein_mut
```
